File: 16. 3Sum Closest/solution.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        """
        Given a list of numbers, nums
        We need to find 3 integers at distinct indices in nums
        Such that their sum either equals or is the closest to 'target'
        """
        nums.sort()
        closest = sum(nums[:3])

        for i in range(len(nums) - 2):
            left = i + 1
            right = len(nums) - 1

            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                if abs(current_sum - target) < abs(closest - target):
                    closest = current_sum

                if current_sum < target:
                    left += 1
                elif current_sum > target:
                    right -= 1
                else:
                    return current_sum

        return closest

File: 16. 3Sum Closest/solution.c
#include <math.h>


void insertion_sort(int arr[], int n)
{
    for (int i = 1; i < n; ++i)
    {
        int key = arr[i];
        int j   = i - 1;

        while ((j >= 0) && (arr[j] > key))
        {
            arr[j + 1] = arr[j];
            j          = j - 1;
        }

        arr[j + 1] = key;
    }
}

/**
 * Given a list of numbers, nums
 * We need to find 3 integers at distinct indices in nums
 * Such that their sum either equals or is the closest to 'target'
 */
int threeSumClosest(int* nums, int numsSize, int target)
{
    insertion_sort(nums, numsSize);
    int closest = nums[0] + nums[1] + nums[2];

    for (int i = 0; i < numsSize - 2; ++i)
    {
        int left  = i + 1;
        int right = numsSize - 1;

        while (left < right)
        {
            int curr_sum = nums[i] + nums[left] + nums[right];
            if (abs(curr_sum - target) < abs(closest - target))
                closest = curr_sum;

            if (curr_sum < target)
                left += 1;
            else if (curr_sum > target)
                right -= 1;
            else
                return curr_sum;
        }
    }

    return closest;
}
